keps/672-serial-job-execution/README.md

```diff
@@ -0,0 +1,499 @@
+ReplicatedJob for pre-trained model and dataset initialization, and the second ReplicatedJob for
+distributed fine-tuning.
+
+The example of JobSet looks as follows:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: fine-tune-llm
+spec:
+  startupPolicy:
+    startupPolicyOrder: InOrder
+    inOrderStartupRules:
+      replicatedJobsWaitForSucceededStatus:
+        - initializer
+      replicatedJobsWaitForReadyStatus: []
+  replicatedJobs:
+    - name: initializer
+      template:
+        spec:
+          template:
+            spec:
+              containers:
+                - name: model-initializer
+                  image: docker.io/kubeflow/model-initializer
+                  volumeMounts:
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+                - name: dataset-initializer
+                  image: docker.io/kubeflow/dataset-initializer
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+    - name: trainer-node
+      template:
+        spec:
+          parallelism: 3
+          completions: 3
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/trainer
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 5
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+```
+
+#### Story 2
+
+As a user, I want to fine-tune my LLM using MPI and DeepSpeed. I have the first
+ReplicatedJob for pre-trained model and dataset initialization, the second ReplicatedJob for MPI launch
+and the third ReplicatedJob for distributed fine-tuning.
+
+The example of JobSet looks as follows:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: deepspeed-mpi
+spec:
+  startupPolicy:
+    startupPolicyOrder: InOrder
+    inOrderStartupRules:
+      replicatedJobsWaitForSucceededStatus:
+        - initializer
+      replicatedJobsWaitForReadyStatus:
+        - launcher
```

Comment:
Wouldn't users have to list here all the replicated jobs that aren't listed in `replicatedJobsWaitForSucceededStatus`?

I would prefer to see something more like this:

```yaml
inOrderStartupRules:
  steps:
  - replicatedJobs: ["initializer"]
  - replicatedJobs: ["launcher", trainer-node"]
```

But names need more work.

Reply:
It could be a good exercise to forget that we ever added `startupPolicyOrder`.

Then, the full API could look something like this:

```yaml
startupPolicy:
  steps:
  - replicatedJobs: ["initializer"]
  - replicatedJobs: ["launcher", "trainer-node"]
    waitForReadiness: ["launcher"]
```

Reply:
>Wouldn't users have to list here all the replicated jobs that aren't listed in replicatedJobsWaitForSucceededStatus?

@alculquicondor No, since we don't need to list `trainer-node` replicatedJob in that list.

>Then, the full API could look something like this:

How can we make sure that steps API are following the order of ReplicatedJobs ?
For example, user can't do this:
```
startupPolicy:
  steps:
  - replicatedJobs: ["launcher"]
  - replicatedJobs: ["initializer", "trainer-node"]
```

Reply:
I am ok with the steps approach, but can one propose a full API please? I don't fully understand the semantics of the proposal in https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1805221427

What we want to capture for each replicatedJob is what status we should be blocking on before proceeding to create the next one, the options are: noWait|Ready|Finished (finished means succeeded or failed)

Reply:
>Clarify something for me: would it be possible to start multiple replicated jobs in parallel?

I think, we haven't discussed it. Basically, what if we want to run Job 1 and Job 2 in parallel before start a Job 3 and Job 4
```
Job 1 -- Succeeded --\
                      Job 3 ---- Ready ---> Job 4
Job 2 -- Succeeded --/
```

@alculquicondor @tenzen-y @ahg-g @kannon92 @danielvegamyhre @vsoch  Do you see an use-case like the above one ?

Reply:
Yes, you'd typically have two models that are going into a higher level analysis. However, I probably wouldn't put this on the level of JobSet but a workflow manager or state machine equivalent.

Reply:
Yeah, I agree that it's worth having such use case in a workflow manager.

But I have heard of customers running workers in two machine types.

So, more like:

```
initializer -> (head, workers type 1 + workers type2)
```

In that case, you might want the head to be ready before the workers, but both types of workers can start together.

IMO, this breaks the original intention of `startupPolicyOrder`. So I would prefer we deprecate it when using "steps" or "barriers"

Reply:
> But I have heard of customers running workers in two machine types.

I think that this is generic use cases. We often use similar machine types (especially GPUs) to improve GPU obtainability, like NVIDIA A100 40GB + A100 80GB (actual distributed Job processes use only 40GB VRAM), when we prioritize GPU obtainability rather than costs (prices).

> IMO, this breaks the original intention of startupPolicyOrder. So I would prefer we deprecate it or not allow to set it when using "steps" or "barriers"

I would recommend deprecating the `startupPolicyOrder` and making `statupPolicyOrder` and `inOrderStartupRules` mutually exclusive, then in the next API version (v1beta1 or something) we remove the `inOrderStartupOrder`.

@andreyvelich @danielvegamyhre @ahg-g What is your opinion?

Reply:
> In that case, you might want the head to be ready before the workers, but both types of workers can start together.

@alculquicondor I think, this use case can be solved with this API:
```
startupPolicy:
  startupPolicyOrder: InOrder
  inOrderStartupRules:
    replicatedJobsWaitForSucceededStatus:
      - initializer
    replicatedJobsWaitForReadyStatus:
      - driver
replicatedJobs:
  - name: initializer
  - name: driver
  - name: trainer-node-nvidia-a100
  - name: trainer-node-nvidia-h100
```


>I would recommend deprecating the startupPolicyOrder and making statupPolicyOrder and inOrderStartupRules mutually exclusive, then in the next API version (v1beta1 or something) we remove the inOrderStartupOrder.

@tenzen-y I think, we should discuss whether we want to support use-case that I introduced above: https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1809273523
For example, if we want to run several Jobs in parallel and wait until they succeeded or become ready.
What if we have TensorFlow distributed training job which has 1 chief, 2 parameter servers, and workers. E.g.:
```
replicatedJobs:
  - name: chief
  - name: ps-1
  - name: ps-2
  - name: trainer-node
```
Do we need to wait until ps-1 and ps-2 is ready before start the trainer-node Job ?

Reply:
>What we want to capture for each replicatedJob is what status we should be blocking on before proceeding to create the next one, the options are: noWait|Ready|Finished (finished means succeeded or failed)

This may not be feasible unless we deprecate the existing startup policy API entirely, but I think an API capturing these semantics could potentially be much simpler and clearer if we embed each ReplicatedJob's dependencies locally into its spec, rather than trying to capture all the dependencies for all replicated jobs in one global JobSet level policy.

For example:

```yaml
replicatedJobs:
  - name: initializer
  - name: driver
     waitForCompleted:
     - initializer
  - name: trainer-node-nvidia-a100
     waitForReady:
     - driver
  - name: trainer-node-nvidia-h100
     waitForReady:
     - driver
```

Instead of:

```yaml
startupPolicy:
  startupPolicyOrder: InOrder
  inOrderStartupRules:
    replicatedJobsWaitForSucceededStatus:
      - initializer
    replicatedJobsWaitForReadyStatus:
      - driver
replicatedJobs:
  - name: initializer
  - name: driver
  - name: trainer-node-nvidia-a100
  - name: trainer-node-nvidia-h100
```

I think packing all these rules into one global policy is going to make the API confusing.

Reply:
I'm still worried this is packing workflow logic into a unit of work, JobSet. It would be akin to kueue not existing and putting scheduling logic all here.

Reply:
>  I think an API capturing these semantics could potentially be much simpler and clearer if we embed each ReplicatedJob's dependencies locally into its spec, rather than trying to capture all the dependencies for all replicated jobs in one global JobSet level policy.

I agree with this. Although having it in a "startupPolicy" field makes it visible at a single glance, as you don't have the podspecs in between. But any option so far has pros and cons.

Going into the details of your proposal, I agree with @vsoch that it might be overlapping too much with a workflow manager. So maybe we can change a little where to put the "hold", like this:

```yaml
replicatedJobs:
  - name: initializer
    waitForStatusBeforeContinuing: Completed
  - name: driver
     waitForStatusBeforeContinuing: Ready
  - name: trainer-node-nvidia-a100
  - name: trainer-node-nvidia-h100
```

Reply:
> @tenzen-y I think, we should discuss whether we want to support use-case that I introduced above: https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1809273523

The JobSet project is not only for ML Traininig, this is used for HPC, CI/CD, and more Job systems.
So, I believe that we should consider the possible use cases during design. Additionally, once we introduce something to the APIs, changing the APIs loses the backward compatibility. So, I would like to evaluate possible designs here.

Reply:
>The JobSet project is not only for ML Traininig, this is used for HPC, CI/CD, and more Job systems.

I think JobSet should focus on ML training and HPC support as the north star - CI/CD systems and other use cases are not really the goal of this project, there's other, more specialized software for that. IMO we should try to do one thing really well, rather than support a bunch of different things less robustly. 

>Do you assume that the waitForStatusBeforeContinuing can be specified only when we specify the startupPolicyOrder: InOrder?

Here we are discussing removing StartupPolicy in favor of embedding the ReplicateJob's dependencies (if any) into the ReplicatedJob spec itself, rather than capturing the dependencies in a global, JobSet level policy, as proposed in my original comment [here](https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1809945813). 

@alculquicondor the change you suggested would be fine with me I think. Curious to get @ahg-g's thoughts as well.

Reply:
> I think JobSet should focus on ML training and HPC support as the north star - CI/CD systems and other use cases are not really the goal of this project

I thought that the JobSet is just a set of Jobs. So I assumed that JobSet covers the user case which Job covers.
But, as you say, JobSet focus on the ML/HPC 

> JobSet: a k8s native API for distributed ML training and HPC workloads

Reply:
> Here we are discussing removing StartupPolicy in favor of embedding the ReplicateJob's dependencies (if any) into the ReplicatedJob spec itself, rather than capturing the dependencies in a global, JobSet level policy, as proposed in my original comment https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1809945813.

I see. In that case, the `AnyOrder` is not supported by the new `waitForStatusBeforeContinuing`, and then the JobSet operator creates Jobs based on the order of replicatedJobs, right?

Reply:
So, after graduating the JobSet API to beta or GA, we will not provide the possibility for the AnyOrder Jobs creation, right?

Reply:
> AnyOrder will still be supported. To achieve AnyOrder startup, the user would simply not specify any dependencies for their ReplicatedJobs (i.e., waitForStatusBeforeContinuing is nil):

That makes sense. Thank you for expanding the API description. +1 on the `waitForStatusBeforeContinuing` approach if @ahg-g does not have any objections.

Reply:
@danielvegamyhre @alculquicondor @tenzen-y  can we support use case that i described here: https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1809570498 If we introduce `waitForStatusBeforeContinuing` ?
E.g. the JobSet should do this:
```
                         PS-1 -- Ready --\
Initializer --Succeeded-->                Trainer
                         PS-2 -- Ready --/
```

Reply:
No, that wouldn't be supported. Is this a common use case or can we apply YAGNI?

Reply:
Actually, it might become common for large scale distributed training use-cases to have 2 parameter servers to reduce load on a single communicator.
I just want to make sure that our API will be future proof.

@alculquicondor @tenzen-y @kannon92 @danielvegamyhre @ahg-g @vsoch  What do you think about this API:
```yaml
replicatedJobs:
  - name: initializer
  - name: ps-1
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: ps-2
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: trainer-node
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
      - name: ps-1
        status: Ready
      - name: ps-2
        status: Ready
```

We can validate that ReplicatedJobs can only reference the previously listed Jobs (e.g. `ps-2` can only reference `ps-1` or `initializer`)

Reply:
Is there no way to represent this more simply - instead of hardcoding "waitForPreviousJobStatusBeforeContinuing" have an generic pattern where there is a rule, and a rule has a trigger and action? For our ensemble-operator (that uses ensemble-python under the hood) I designed a pretty simple pattern, and it's event driven based on events from the workload manager queue ([example](https://github.com/converged-computing/ensemble-python/blob/80f5c095c4483d5e29402097443eb0dedbab62cc/examples/grow-shrink-example.yaml#L21-L41)). Could a replicatedJob not have a similar list of rules that respond to events (or conditions) with some action?

Reply:
I might be very hard to manage quota for such Job sequence with rules.
E.g. Job dispatcher needs to analyse set of rules for each ReplicatedJob to understand how to allocate quota.

@vsoch Do you have any JobSet API example in mind that we can discuss here ?

Reply:
I was planning on doing exactly that, but JobSet is immutable in every way so I couldn’t put it on this context. My plan is to put together a demo/talk for the group to teach about ensembles and the use case, and then hopefully be able to provide that.

Reply:
> to have 2 parameter servers to reduce load on a single communicator.

Wouldn't those be represented as two jobs in a single replicatedJob?

I don't want to dismiss the use case, but also the API described in https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1818283828 starts to become complicated to the most common cases where this will be used.

If we consider https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1814857024 as out of scope, I feel the API currently described in the KEP is simple and it is able to clearly describe the targeted use cases, or did I miss something?

Reply:
> to have 2 parameter servers to reduce load on a single communicator.
>> Wouldn't those be represented as two jobs in a single replicatedJob?

I think @ahg-g is correct about this.

>If we consider https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1814857024 as out of scope, I feel the API currently described in the KEP is simple and it is able to clearly describe the targeted use cases, or did I miss something?

@ahg-g I agree we should consider that out of scope. However, I actually think the API currently described in the KEP is more complicated, and the alternative above is a better fit. This thread is getting long so let's do a side-by-side comparison here.

Option 1: JobSet level API currently proposed in the KEP (global set of startup policy rules which capture the local dependencies of all replicatedJobs): 
  ```yaml
  
   startupPolicy:
    startupPolicyOrder: InOrder
    inOrderStartupRules:
      replicatedJobsWaitingOnReadyStatus:
      - job-2
      replicatedJobsWaitingOnSucceededStatus:
      - job-1
  replicatedJobs:
    - name: job-1
    - name: job-2
    - name: job-3
```

Option 2: ReplicatedJob level API (each replicatedJob's local condition to continue is embedded in the replicatedJob spec itself):

```yaml
replicatedJobs:
  - name: initializer
    waitForStatusBeforeContinuing: Completed
  - name: driver
    waitForStatusBeforeContinuing: Ready
  - name: trainer-node-nvidia-a100
  - name: trainer-node-nvidia-h100
  ```

Personally I think the status that a given replicatedJob must reach before the subsequent replicatedJob begins execution is an inherently local attribute of that replicatedJob, so it is a natural fit for an API local to the replicatedJob (option 2). 

In addition, the global rules-based API is separating the startup order (as defined in the `replicatedJobs[]`order) from the startup rules (as defined in the list of `inOrderStartupPolicyRules[]`), when I think they really belong together. Separating them requires the user to repeatedly cross-reference different parts of the manifest to understand the full execution model of the JobSet. This becomes especially cumbersome for larger, more complicated JobSets.

In contrast, the local API captures both startup order and startup rules in `replicatedJobs[]`, so you can simply read the manifest from top to bottom in an intuitive fashion.

In my opinion, the current API described in the KEP feels a bit like we are trying force something with inherently local semantics into a global, rules-based API because it more closely follows the same rules-based pattern we have in other parts of the JobSet API (failure policy rules, etc) and other upstream k8s APIs (pod failure policy rules, etc). Having said that, the consistency with other APIs is an attractive property, so there is a trade-off.

@ahg-g @andreyvelich what are your thoughts on this?

Reply:
I think for the most part the number of replicated jobs will be limited, and so expressing the policy in one place should not be too bad.

> In contrast, the local API captures both startup order and startup rules in replicatedJobs[], so you can simply read the manifest from top to bottom in an intuitive fashion.

It is spread out though with full job/pod specs in between.

I see both approaches to have pros/cons, but I lean towards the existing format because it expresses the policy in one place. But does it express everything we want? @andreyvelich?

Reply:
>Wouldn't those be represented as two jobs in a single replicatedJob?

Yes, that could work as long as we don't want to have 2 different PodSpec templates for those Parameter Servers. For example, to use different Node affinity. Since ReplicatedJob only has single PodTemplateSpec.

>@ahg-g @andreyvelich what are your thoughts on this?

Personally, I prefer the Option 2. Given that the FailurePolicy and SuccessPolicy has a different meaning and we don't need to define FailurePolicy for every ReplicatedJob.

@danielvegamyhre @ahg-g What do we think about API that I proposed here: https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1818283828?

Reply:
> @danielvegamyhre @ahg-g What do we think about API that I proposed here: [#680 (comment)](https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1818283828)?

What is the behavior for the setups?

Setup 1
```
replicatedJobs:
  - name: initializer
  - name: ps-1
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: ps-2
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: trainer-node
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Ready            <------------ changed this from Succeeded to Ready
      - name: ps-1
        status: Ready
      - name: ps-2
        status: Ready
```

Setup 2

```
replicatedJobs:
  - name: initializer
  - name: ps-1
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: ps-2
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: trainer-node
    waitForPreviousJobStatusBeforeContinuing:   
      - name: ps-2                     <------------ removed the first two items
        status: Ready
```

Setup 3

```
replicatedJobs:
  - name: initializer
  - name: ps-1
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: ps-2
    waitForPreviousJobStatusBeforeContinuing:
      - name: initializer
        status: Succeeded
  - name: trainer-node. <------------ removed all items
```

Reply:
I am concerned that this will not be easy to reason about compared to simple sequences. I am still not sure we need this added complexity. Can we have the API initially support strict sequences and extend it in the future to support this case? Basically not have the list of replicatedJobs and assume that the waitForPreviousJobStatusBeforeContinuing refers to the previous replicatedJob.

Reply:
Sure, I am ok with that, so we just not introduce `name` parameter into API, right ?
Do you prefer Option 1 or Option 2 ?
Option 1
```
replicatedJobs:
  - name: initializer
  - name: ps-1
    waitForPreviousJobStatusBeforeContinuing:
      status: Completed
  - name: trainer-node
    waitForPreviousJobStatusBeforeContinuing:
      status: Ready
```

Option 2

```
replicatedJobs:
  - name: initializer
    waitForStatusBeforeContinuing: Completed
  - name: ps-1
    waitForStatusBeforeContinuing: Ready
  - name:  trainer-node
```

Reply:
@ahg-g I think, it would be breaking change, since for Option 1, we should add the slice:
```
  - name: ps-1
    waitForPreviousJobStatusBeforeContinuing:
      - status: Completed
        name: initializer
```
For option 2, I don't think we can extend it.